### PR TITLE
Upgrade to LLVM7 and Clang7

### PR DESCRIPTION
**Pull request by FabianWiebe**
### Main Changes
This pr updates LLVM6 and Clang6 to LLVM7 and Clang7.
This requires the replacement of the LLVM6 files `CloneFunction.cpp` and `InlineFunction.cpp` with their LLVM7 pendant required for JIT.
The interface to pass a module to the JIT compiler slightly changed so that modules are now stored in a unique_ptr instead of a shared_ptr.

### Support for LLVM7 and Clang7
MacOS - Brew will automatically install the latest version of LLVM with `brew install llvm`.
Ubuntu 18.10 supports LLVM7 and Clang7.
Ubuntu 18.04 does not support LLVM7 and Clang7 out of the box.
To install the required versions in Ubuntu 18.04, add the following lines to `/etc/apt/sources.list`.
````
deb http://apt.llvm.org/stretch/ llvm-toolchain-stretch-7 main
# deb-src http://apt.llvm.org/stretch/ llvm-toolchain-stretch-7 main
````
PR removes one issue from #1077.
#### Open issues:
- remove workaround to remove LLVM lib files from coverage, see #1074
- remove experimental/filesystem (reopen #1151)

### CODE REVIEW

**Comment by FabianWiebe** [comment]
Once the Jenkins supports LLVM7 and Clang7, the tests should pass.

**Comment by mrzzzrm** [comment]
Maybe add "Fixes #1077" to the OP so this PR closes the issue automatically?

**Comment by Bouncner** [comment]
So the "further steps" will not be part of this PR? I am confused.

**Comment by FabianWiebe** [comment]
I renamed `further steps` to `open issues`. These issues could be integrated in this pr but could also be incorporated through separate ones. For #1151 exists already a closed pr which could be reopened. So far, I did not look into those two issues so that I would suggest that they are addressed by someone already familiar with them.

File: src/plugins/CMakeLists.txt [hunk 1]
```diff
@@ -24,8 +24,8 @@ function(add_plugin)
     endif()
 endfunction(add_plugin)
 
-add_plugin(NAME TestPlugin SRCS test_plugin.cpp test_plugin.hpp)
-add_plugin(NAME TestNonInstantiablePlugin SRCS non_instantiable_plugin.cpp)
+add_plugin(NAME hyriseTestPlugin SRCS test_plugin.cpp test_plugin.hpp)
```
Comment (mrks) [inline src/plugins/CMakeLists.txt]: @Bensk1 llvm brings its own "TestPlugin", so I had to rename this one.

**Comment by mrks** [comment]
Looks like clang is broken again. I first thought this was fixed by [r333715](https://github.com/llvm/llvm-project/commit/d781d97ed45259809bc19d97a77fde508e2a036d#diff-9a2cfc9e3da40490bc37dc746ec36963), but that is already part of llvmorg-7.0.0. It is, however, fixed in my master build of clang-8.

**Comment by mrks** [comment]
This is fixed in clang 7.0.1, which was released in December but has not made it into ubuntu's default packages.

@FabianWiebe: Could you address the clang-tidy issues, please? Once that is done, I'd suggest we wait for the packages to be updated.

**Comment by mrks** [comment]
It might be easier to create your own, local clang-tidy folder. Have a look at the Jenkinsfile to see the cmake options for that.

**Comment by mrks** [comment]
Ah crap. I didn't realize until now that you would face clang-tidy issues that don't even come from the JIT code. Let me know if that becomes too much and makes it difficult for you to get the results of your thesis in.

File: src/benchmarklib/table_generator.cpp [hunk 2]
```diff
@@ -100,7 +101,7 @@ std::shared_ptr<Table> TableGenerator::generate_table(
 
   std::random_device rd;
   // using mt19937 because std::default_random engine is not guaranteed to be a sensible default
-  auto pseudorandom_engine = std::mt19937{};
+  auto pseudorandom_engine = std::mt19937{};  // NOLINT(cert-msc32-c, cert-msc51-cpp)
```
Comment (mrks) [inline src/benchmarklib/table_generator.cpp]: should we remove those two checks from the clang-tidy list?

File: src/bin/console/console.cpp [hunk 3]
```diff
@@ -1035,6 +1035,7 @@ bool Console::_handle_rollback() {
 
 }  // namespace opossum
 
+// NOLINTNEXTLINE(bugprone-exception-escape) - clang-tidy does not like exceptions in function 'main'
```
Comment (mrks) [inline src/bin/console/console.cpp]: Remove check from tidy?

File: src/lib/concurrency/transaction_context.cpp [hunk 4]
```diff
@@ -16,6 +16,7 @@ TransactionContext::TransactionContext(const TransactionID transaction_id, const
       _phase{TransactionPhase::Active},
       _num_active_operators{0} {}
 
+// NOLINTNEXTLINE(bugprone-exception-escape) - clang-tidy does not like exceptions in destructors
```
Comment (mrks) [inline src/lib/concurrency/transaction_context.cpp]: Remove check from tidy?

File: src/lib/operators/aggregate.cpp [hunk 5]
```diff
@@ -123,6 +123,7 @@ template <typename ColumnDataType, typename AggregateType, typename AggregateKey
 struct AggregateContext : public AggregateResultContext<ColumnDataType, AggregateType> {
   AggregateContext() {
     auto allocator = AggregateResultIdMapAllocator<AggregateKey>{&this->buffer};
+    // NOLINTNEXTLINE(clang-analyzer-core.CallAndMessage) - false warning: called C++ object (result_ids) is null
```
Comment (mrks) [inline src/lib/operators/aggregate.cpp]: Maybe add `: result_ids(nullptr)` instead?

File: src/lib/statistics/chunk_statistics/histograms/equal_width_histogram.cpp [hunk 6]
```diff
@@ -261,6 +261,7 @@ BinID EqualWidthHistogram<T>::_bin_for_value(const T& value) const {
     // The exception is if this is the last bin, then it is actually part of the last bin,
     // because that edge is stored separately and therefore not trimmed to the prefix length.
     // We checked earlier that it is not larger than maximum().
+    // NOLINTNEXTLINE(abseil-string-find-startswith) - false warning: absl::StartsWith only available in C++20
```
Comment (mrks) [inline src/lib/statistics/chunk_statistics/histograms/equal_width_histogram.cpp]: That's a nice warning though.

File: src/lib/operators/aggregate.cpp (hunk 5 above)
Comment (FabianWiebe) [reply src/lib/operators/aggregate.cpp]: This does not resolve the warning.
`: result_ids{std::make_unique<AggregateResultIdMap<AggregateKey>>()}`
or
`: result_ids{nullptr}`
still result in a warning:
````
/hyrise/src/lib/operators/aggregate.cpp:127:5: error: Called C++ object pointer is null [clang-analyzer-core.CallAndMessage,-warnings-as-errors]
    result_ids = std::make_unique<AggregateResultIdMap<AggregateKey>>(allocator);
    ^
````

Comment (mrks) [reply src/lib/operators/aggregate.cpp]: Ok, I don't understand its problem.

**Comment by mrks** [comment]
Clang 7.0.1 will not find its way into Ubuntu 18.10. Let's close this for now and wait for 19.04. #1077 will make sure that it does not get lost.

**Comment by mrks** [comment]
(18.10 remains unchanged on 7.0.0-3)

**Pull request closed without merging**